File: InitialProofOfConcept/InitialProofOfConcept/AKFMOscillator.m
//
//  AKFMOscillator.m
//  AudioKit
//
//  Created by Aurelius Prochazka on 9/3/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "AKFMOscillator.h"

@implementation AKFMOscillator
{
    AKManager *manager;
    AKTable *_waveform;
    float out;
}

- (instancetype)init
{
    NSLog(@"initing fmoscillator");
    if ([super init]) {
        manager = [AKManager sharedManager];
        [self create];
    }
    return self;
    
}

- (instancetype)initWithWaveform:(AKTable *)waveform
                    baseFrequency:(AKParameter *)baseFrequency
                carrierMultiplier:(AKParameter *)carrierMultiplier
             modulatingMultiplier:(AKParameter *)modulatingMultiplier
                  modulationIndex:(AKParameter *)modulationIndex
                        amplitude:(AKParameter *)amplitude
{
    self = [super init];
    if (self) {
        manager = [AKManager sharedManager];
        _waveform = waveform;
        [self create];
        [baseFrequency bind:&_fosc->freq];
        [carrierMultiplier bind:&_fosc->car];
        [modulatingMultiplier bind:&_fosc->mod];
        [modulationIndex bind:&_fosc->indx];
        [amplitude bind:&_fosc->amp];
    }
    return self;
}

- (instancetype)initWithBaseFrequency:(AKParameter *)baseFrequency
                    carrierMultiplier:(AKParameter *)carrierMultiplier
                 modulatingMultiplier:(AKParameter *)modulatingMultiplier
                      modulationIndex:(AKParameter *)modulationIndex
                            amplitude:(AKParameter *)amplitude
{
    AKTable *waveform = [[AKTable alloc] init];
    return [self initWithWaveform:waveform
                    baseFrequency:baseFrequency
                carrierMultiplier:carrierMultiplier
             modulatingMultiplier:modulatingMultiplier
                  modulationIndex:modulationIndex
                        amplitude:amplitude];
}

- (void)create
{
    sp_fosc_create(&_fosc);
    sp_fosc_init(manager.data, _fosc, _waveform.table);
    out = 0;
// This could change..
    self.value = &out;
}

- (float)compute
{
    sp_fosc_compute(manager.data, _fosc, NULL, &out);
    *self.value = out;
    return out;
}

- (void)destroy
{
    [_waveform destroy];
    sp_fosc_destroy(&_fosc);
    [manager destroy];
}

@end

File: SwiftOSXProofOfConcept/SwiftOSXProofOfConcept/AKFMOscillator.swift
//
//  AKFMOscillator.swift
//  AudioKit
//
//  Created by Aurelius Prochazka on 9/5/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

import Foundation

/** Frequency Modulating Oscillator */
@objc class AKFMOscillator : AKParameter {
    
    var fosc = UnsafeMutablePointer<sp_fosc>.alloc(1) // allocate 1
    
    var table = AKTable()

    var frequency: Float = 440 {
        didSet {
            fosc.memory.freq = frequency
        }
    }
    
    var index: Float = 1 {
        didSet {
            fosc.memory.indx = index
        }
    }
    
    /** Instantiates the fm oscillator with default values */
    override init() {
        super.init()
        create()
    }
    
    /** Instantiates the fm oscillator with all values
    
    :param: baseFrequency In cycles per second, or Hz, this is the common denominator for the carrier and modulating frequencies. Updated at Control-rate. [Default Value: 440]
    :param: carrierMultiplier This multiplied by the baseFrequency gives the carrier frequency. [Default Value: 1]
    :param: modulatingMultiplier This multiplied by the baseFrequency gives the modulating frequency. [Default Value: 1]
    :param: modulationIndex This multiplied by the modulating frequency gives the modulation amplitude. Updated at Control-rate. [Default Value: 1]
    :param: amplitude This multiplied by the modulating frequency gives the modulation amplitude. [Default Value: 0.5]
    */
    convenience init(
        baseFrequency: AKParameter,
        carrierMultiplier: AKParameter,
        modulatingMultiplier: AKParameter,
        modulationIndex: AKParameter,
        amplitude: AKParameter)
    {
        self.init()
        baseFrequency.bind(&fosc.memory.freq)
        carrierMultiplier.bind(&fosc.memory.car)
        modulatingMultiplier.bind(&fosc.memory.mod)
        modulationIndex.bind(&fosc.memory.indx)
        amplitude.bind(&fosc.memory.amp)
    }
    
    override func create() {
    /** Internal set up function */
        sp_fosc_create(&fosc)
        sp_fosc_init(AKManager.sharedManager.data, fosc, table.ftbl)
    }
    
    /** Computation of the next value */
    override func compute() -> Float {
        sp_fosc_compute(AKManager.sharedManager.data, fosc, nil, &value);
        pointer.memory = value
        return value
    }
    
    override func destroy() {
    /** Release of memory */
        sp_fosc_destroy(&fosc)
    }
    
    
}